### PR TITLE
[ADD] l10n_in_gstin_status: check GSTIN status

### PR DESCRIPTION
**Pull request by hamo-odoo**
This commit introduces a new module enabling users to check the GSTIN status of vendors. We can install it manually or with 'GST Status API Service' in the 'Customer Invoices' section in the Accounting settings.

After installation, Users will find a 'Check Status'  button on the partner form. By clicking, the system retrieves the latest GSTIN status from the service provider, displaying the GSTIN status, and the date of the last verification. Now the 'Re Verify' button is shown on partner form, vendor bills, customer invoices, credit notes, and debit notes to re verify the status. This enhancement ensures users have up-to-date information regarding vendor compliance directly within Odoo.

task-3707122

see https://github.com/odoo/enterprise/pull/67787
see https://github.com/odoo/upgrade/pull/6335

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_in_gstin_status/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            if date_from and re.search(r'\d', date_from):
+                message = _(
+                    "GSTIN %(vat)s is %(status)s and Effective from %(date_from)s",
+                    vat=self.vat,
+                    status=gst_status,
+                    date_from=date_from,
+                )
+            else:
+                message = _(
+                    "GSTIN %(vat)s is %(status)s, but the response does not provide the effective date.",
```
Comment (jva-odoo) [inline addons/l10n_in_gstin_status/models/res_partner.py:86]:
```suggestion
                    "GSTIN %(vat)s is %(status)s, but effective date is not available.",
```

File: addons/l10n_in_gstin_status/tests/test_check_status.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                "error": [{"code": "SWEB_9035", "message": "Invalid GSTIN / UID"}],
+            },
+            "no_records": {
+                "error": [{"code": "FO8000", "message": "No records found for the provided GSTIN."}],
+            },
+        }
+
+    @freeze_time('2024-05-20')
+    def check_gstin_status(self, partner, expected_status, mock_response, raises_exception=False):
+        with patch("odoo.addons.l10n_in_check_gst_status.models.res_partner.jsonrpc") as mock_jsonrpc:
```
Comment (jva-odoo) [inline addons/l10n_in_gstin_status/tests/test_check_status.py:50]:
```suggestion
        with patch("odoo.addons.l10n_in_gstin_status.models.res_partner.jsonrpc") as mock_jsonrpc:
```

File: addons/l10n_in_gstin_status/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            if partner.country_code == 'IN':
+                partner.l10n_in_gstin_verified_status = False
+                partner.l10n_in_gstin_verified_date = False
+
+    def _l10n_in_gstin_status_get_url(self, company):
+        if company.sudo().l10n_in_edi_production_env:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_check_gst_status.endpoint", default_endpoint)
```
Comment (jva-odoo) [inline addons/l10n_in_gstin_status/models/res_partner.py:45]:
```suggestion
        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_gstin_status.endpoint", default_endpoint)
```

**Review by jva-odoo — COMMENTED** [review]
else okay

**Review by jva-odoo — APPROVED** [review]
@jco-odoo LGTM

**Comment by jco-odoo** [comment]
@jva-odoo Do not we need an upgrade script for this one as well?

**Comment by jva-odoo** [comment]
@jco-odoo I don't think we need migration for this because we do not remove the field but just move from one module to another one.
But I am not sure so we need to confirm this

**Comment by jco-odoo** [comment]
@jva-odoo util.move_field_to_module 
And maybe we are merging two fields?

**Comment by jva-odoo** [comment]
And move to the higher module as per dependency.
so first `l10n_in` is updated and after `l10n_in_edi`,  ORM can handle this as per my knowledge

**Comment by jva-odoo** [comment]
Okay, we do migration for this.
sorry for misunderstanding

File: addons/l10n_in_gstin_status/views/res_partner_base_vat_views.xml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            <field name="inherit_id" ref="base_vat.view_partner_base_vat_form" />
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='vies_valid']" position="after">
+                    <span invisible="country_code != 'IN' or not vat">
+                        <span invisible="not l10n_in_gstin_verified_date or not l10n_in_gstin_verified_status"
+                            class="oe_inline text-success">Active</span>
+                        <span invisible="not l10n_in_gstin_verified_date or l10n_in_gstin_verified_status"
+                            class="oe_inline text-danger">Inactive</span>
+                        <span invisible="not l10n_in_gstin_verified_date and not l10n_in_gstin_verified_status" class="text-muted">
+                            <span class="ps-2">checked: </span>
```
Comment (jco-odoo) [inline addons/l10n_in_gstin_status/views/res_partner_base_vat_views.xml:16]:
```suggestion
 (
```
If you have "Active (02/03/2025)", would not it be as clear?

File: addons/l10n_in_gstin_status/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            if partner.country_code == 'IN':
+                partner.l10n_in_gstin_verified_status = False
+                partner.l10n_in_gstin_verified_date = False
+
+    def _l10n_in_gstin_status_get_url(self, company):
+        if company.sudo().l10n_in_edi_production_env:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_gstin_status.endpoint", default_endpoint)
```
Comment (jco-odoo) [inline addons/l10n_in_gstin_status/models/res_partner.py:45]: a different parameter here as well?  Apart from the /iap/l10n_in_reports/1/public/search, this method could be in l10n_in

File: addons/l10n_in/models/company.py [hunk 6]
```diff
@@ -4,6 +4,12 @@
 class ResCompany(models.Model):
     _inherit = 'res.company'
 
+    l10n_in_edi_production_env = fields.Boolean(
+        string="Indian Electronic Processing Is production",
+        help="Enable the use of production credentials",
+        groups="base.group_system",
+        default=True
```
Comment (jco-odoo) [inline addons/l10n_in/models/company.py:11]:
```suggestion
```
We did not have the default production to True before and it is better to let people explicitly say it is going to official to avoid any accidents. Or what is the reason for this change?

File: addons/l10n_in_gstin_status/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            else:
+                message = _(
+                    "GSTIN %(vat)s is %(status)s, effective date is not available.",
+                    vat=self.vat,
+                    status=gst_status
+                )
+            self.message_post(body=message)
+        else:
+            _logger.info("GST status check error %s", response)
+            raise UserError(_("Error in getting GST status. please try again"))
```
Comment (jco-odoo) [inline addons/l10n_in_gstin_status/models/res_partner.py:93]: I have been getting this all the time checking on runbot https://66355233-master-all.runbot164.odoo.com/odoo/customers/147

and it is not very helpful...   It is like: we get an error and we do not know why it is there. (but it could connect)  We should be more careful as to why this happens.  (as this case should not even happen)

Comment (jco-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py:93]: Checking in the logs of the runbot, we get this: 
2024-08-02 08:35:05,297 29 INFO 66355233-master-all odoo.addons.l10n_in_gstin_status.models.res_partner: GST status check error {'error': [{'code': 'SWEB_9035', 'message': 'Invalid GSTIN / UID'}]} 
See https://runbot164.odoo.com/runbot/static/build/66355233-master/logs/run.txt

We should explicitly say that the GSTIN is invalid.  (and not let the user retry)

File: addons/l10n_in_gstin_status/views/res_partner_base_vat_views.xml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jco-odoo) [inline addons/l10n_in_gstin_status/views/res_partner_base_vat_views.xml:1]: The name of this file is a bit long.  res_partner_views.xml seems clear enough.

File: addons/l10n_in/models/company.py (hunk 6 above)
Comment (jva-odoo) [reply addons/l10n_in/models/company.py:11]: The reason is too many tickets related to this "users don't tick this and try to validate credentials."
And so the call goes to the testing server and it's failed to validate.

Also, this new service does not work without production.
And when we create a database with demo data then we False this for that company.

Comment (jco-odoo) [reply addons/l10n_in/models/company.py:11]: Is not there something we can do usability wise to make them not forget to check it?

Comment (jva-odoo) [reply addons/l10n_in/models/company.py:11]: It's not a big deal we can improve it later

File: addons/l10n_in/const.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+DEFAULT_IAP_ENDPOINT = "https://l10n-in-edi.api.odoo.com"
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/const.py:1]: As I ref other modules we set URLs in const.py so still I suggest not moving the method in res_company

File: addons/l10n_in_gstin_status/models/res_partner.py (hunk 5 above)
Comment (hamo-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py]: > a different parameter here as well? Apart from the /iap/l10n_in_reports/1/public/search, this method could be in l10n_in

I did not get this comment rest have been applied

File: addons/l10n_in/const.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hamo-odoo) [reply addons/l10n_in/const.py:1]: I noticed that the `payment_*` modules include a const file. Since URLs are being used across other modules, wouldn't it be more consistent and efficient for the URLs to be defined in the constants file rather than within model?

**Comment by hamo-odoo** [comment]
@jco-odoo Done with the changes.

File: addons/l10n_in_gstin_status/models/res_partner.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                )
+            else:
+                message = _(
+                    "GSTIN %(vat)s is %(status)s, effective date is not available.",
+                    vat=self.vat,
+                    status=gst_status
+                )
+            self.message_post(body=message)
+        else:
+            _logger.info("GST status check error %s", response)
```
Comment (jco-odoo) [inline addons/l10n_in_gstin_status/models/res_partner.py:84]: image

Should not we do a separate message for when the GST is invalid?  @hamo-odoo

Comment (jva-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py:84]: @jco-odoo Or just an error message and code
like `[SWEB_9035] Invalid GSTIN/UIN`

Comment (jva-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py:84]: if no error message and code then
`Unknown error while checking GST status`

Comment (jco-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py:84]: I still think the invalid GSTIN would deserve its own message as it could be a pretty common case, but the way you suggest is already way better.

Comment (jva-odoo) [reply addons/l10n_in_gstin_status/models/res_partner.py:84]: @jco-odoo okay,
@hamo-odoo wrote a good message for a special code `SWEB_9035` we get the benefit of translation in this case.

**Comment by hamo-odoo** [comment]
@jva-odoo @jco-odoo Changes are done. As per JVA suggestion the error message from API itself is good enough for user. I have made changes accordingly adding a default error message. Also added specific error message for error code `SWEB_9035`

File: addons/l10n_in_edi/models/account_edi_format.py [hunk 11]
```diff
@@ -650,21 +647,17 @@ def _l10n_in_edi_get_token(self, company):
...
         try:
-            return jsonrpc(url, params=params, timeout=25)
+            return self.env['iap.account']._l10n_in_connect_to_server(
+              company.sudo().l10n_in_edi_production_env,
+              params,
+              url_path,
+              "l10n_in_edi.endpoint"
+            )
         except AccessError as e:
             _logger.warning("Connection error: %s", e.args[0])
```
Comment (jco-odoo) [inline addons/l10n_in_edi/models/account_edi_format.py:662]: We should think to improve: what if not enough credits?  https://github.com/odoo/odoo/pull/175290/files#diff-8c0bdd04813477607a9aeeab7bbb952a1a22303ae588a8196f44343fcb3af89dR52

File: addons/l10n_in/views/res_config_settings_views.xml [hunk 12]
```diff
@@ -9,17 +9,49 @@
...
-                </setting>
             </block>
             <xpath expr="//app[@name='account']" position="inside">
-                <div id="india_integration_section" invisible="1">
+                <div id="india_integration_section">
                     <block title="Indian Integration" id="india_localization" invisible="country_code != 'IN'">
+                        <setting name="india_production_setting"
+                                 string="Production Environment"
+                                 help="Activate this to start using Indian services in the production environment."
+                                 >
```
Comment (jco-odoo) [inline addons/l10n_in/views/res_config_settings_views.xml:19]:
```suggestion
                                 help="Activate this to start using Indian services in the production environment.">
```

**Comment by jco-odoo** [comment]
@robodoo r+
